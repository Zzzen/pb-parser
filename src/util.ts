export function assertIsDefined<T>(
  value: T,
  message: string
): asserts value is NonNullable<T> {
  // eslint-disable-next-line no-null/no-null
  if (value === undefined || value === null) {
    throw new Error(message);
  }
}
