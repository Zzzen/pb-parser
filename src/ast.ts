import { Token } from "./scanner";

export interface Comment {
  type: "Comment";
  value: string;
  start: number;
  end: number;
}

export interface Location {
  // starts from 1
  line: number;
  // starts from 0
  column: number;
}

export interface WithFullLocation {
  start: number;
  end: number;
  loc: {
    start: Location;
    end: Location;
  };
}

export interface BaseNode extends WithFullLocation {
  leadingComments?: Comment[];
  trailingComments?: Comment[];
  startToken: Token;
  endToken: Token;
}

export interface SyntaxStatement extends BaseNode {
  type: "SyntaxStatement";
  value: "proto2" | "proto3";
}

export interface ImportStatement extends BaseNode {
  type: "ImportStatement";
  file: string;
  modifier: "weak" | "public";
}

export type Label = "required" | "optional" | "repeated";

export interface Option extends BaseNode {
  type: "Option";
  name: OptionName;
  value: Constant;
}

export interface ProtoFile extends BaseNode {
  type: "ProtoFile";
  syntax?: SyntaxStatement;
  body: TopLevelDirective[];
  tokens: Token[];
}

export type TopLevelDirective =
  | ImportStatement
  | Package
  | Option
  | TopLevelDef
  | EmptyStatement;

export type TopLevelDef = Message | Enum | Extend | Service;

export interface Package extends BaseNode {
  type: "Package";
  name: FullIdentifier;
}

export interface FullIdentifier extends BaseNode {
  type: "FullIdentifier";
  value: string;
}

export interface OptionName extends BaseNode {
  type: "OptionName";
  value: string;
}

export type Constant = FullIdentifier | Literal;

export interface Literal extends BaseNode {
  type: "Literal";
  value: string | number | boolean;
}

export interface Message extends BaseNode {
  type: "Message";
  name: string;
  body: MessageBody[];
}

export type MessageBody =
  | Field
  | Enum
  | Message
  | Extend
  | Extensions
  | Group
  | Option
  | OneOf
  | MapField
  | Reserved;

export interface Enum extends BaseNode {
  type: "Enum";
  name: string;
  options: Option[];
  fields: Field[];
}

export interface Extend extends BaseNode {
  type: "Extend";
}

export interface Service extends BaseNode {
  type: "Service";
}

export interface Field extends BaseNode {
  type: "Field";
  name: string;
  label?: "repeated" | "optional" | "required";
  fieldType?: ValueType;
  fieldNumber: number;
  fieldOptions: Option[];
}

export interface Extensions extends BaseNode {
  type: "Extensions";
  ranges: Array<[number, number | "max"] | [number]> | undefined;
}

export interface Group extends BaseNode {
  type: "Group";
  label: Label;
  groupName: string;
  fieldNumber: number;
  body: MessageBody[];
}

export interface OneOf extends BaseNode {
  type: "OneOf";
  oneofName: string;
  options: Option[];
  fields: Field[];
}

export interface MapField extends BaseNode {
  type: "MapField";
  keyType: string;
  valueType: ValueType;
  fieldNumber: number;
  mapName: string;
  options: Option[];
}

export interface ValueType extends BaseNode {
  type: "ValueType";
  value: string;
}

export interface Reserved extends BaseNode {
  type: "Reserved";
  ranges: Array<[number, number | "max"] | [number]> | undefined;
  fieldName: string[] | undefined;
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EnumField extends BaseNode {
  type: "EnumField";
}

export interface EnumValueOption extends BaseNode {
  type: "EnumValueOption";
}

export interface EnumBody extends BaseNode {
  type: "EnumBody";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}

export interface EmptyStatement extends BaseNode {
  type: "EmptyStatement";
}
