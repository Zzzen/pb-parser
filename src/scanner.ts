import { Location } from "./ast";
import { last } from "./util";

export enum TokenType {
  // Single-character tokens.
  LEFT_PAREN = "LEFT_PAREN",
  RIGHT_PAREN = "RIGHT_PAREN",
  LEFT_BRACE = "LEFT_BRACE",
  RIGHT_BRACE = "RIGHT_BRACE",
  LEFT_BRACKET = "LEFT_BRACKET",
  RIGHT_BRACKET = "RIGHT_BRACKET",
  COMMA = "COMMA",
  DOT = "DOT",
  MINUS = "MINUS",
  PLUS = "PLUS",
  SEMICOLON = "SEMICOLON",
  SLASH = "SLASH",
  STAR = "STAR",
  QUESTION_MARK = "QUESTION_MARK",
  COLON = "COLON",
  COMMENT = "COMMENT",

  // One or two character tokens.
  BANG = "BANG",
  BANG_EQUAL = "BANG_EQUAL",
  EQUAL = "EQUAL",
  EQUAL_EQUAL = "EQUAL_EQUAL",
  GREATER = "GREATER",
  GREATER_EQUAL = "GREATER_EQUAL",
  LESS = "LESS",
  LESS_EQUAL = "LESS_EQUAL",

  // Literals.
  IDENTIFIER = "IDENTIFIER",
  STRING = "STRING",
  NUMBER = "NUMBER",

  // int32, int64, string, bytes...
  PRIMITIVE_TYPE = "PRIMITIVE_TYPE",

  EOF = "EOF",
}

const PrimitiveTypes = [
  "int32",
  "int64",
  "uint32",
  "uint64",
  "sint32",
  "sint64",
  "fixed32",
  "fixed64",
  "sfixed32",
  "sfixed64",
  "bool",
  "string",
  "bytes",
] as const;

/**
 * keyword is a subset of IDENTIFIER
 */
export enum Keyword {
  SYNTAX = "syntax",
  IMPORT = "import",
  WEAK = "weak",
  PUBLIC = "public",
  PACKAGE = "package",
  OPTION = "option",
  REQUIRED = "required",
  OPTIONAL = "optional",
  REPEATED = "repeated",
  ONEOF = "oneof",
  GROUP = "group",
  MAP = "map",
  EXTENSIONS = "extensions",
  TO = "to",
  RESERVED = "reserved",
  ENUM = "enum",
  MESSAGE = "message",
  EXTEND = "extend",
  SERVICE = "service",
  RPC = "rpc",
  RETURNS = "returns",
  MAX = "max",
}

export interface Token {
  type: TokenType;
  lexeme: string;
  literal?: unknown;
  start: number;
  end: number;
  loc: {
    start: Location;
    end: Location;
  };
}

const base10Re = /^[1-9][0-9]*/;
// const base10NegRe = /^-?[1-9][0-9]*/
const base16Re = /^0[xX][0-9a-fA-F]+/;
// const base16NegRe = /^-?0[x][0-9a-fA-F]+/
const base8Re = /^0[0-7]+/;
// const base8NegRe = /^-?0[0-7]+/
const numberRe = /^(?![eE])[0-9]*(?:\.[0-9]*)?(?:[eE][+-]?[0-9]+)?/;
const nameRe = /^[a-zA-Z_][a-zA-Z_0-9]*/;
// const typeRefRe = /^(?:\.?[a-zA-Z_][a-zA-Z_0-9]*)(?:\.[a-zA-Z_][a-zA-Z_0-9]*)*/;
// const fqTypeRefRe = /^(?:\.[a-zA-Z_][a-zA-Z_0-9]*)+/;

export class Scanner {
  current = 0;
  start = 0;
  line = 0;
  tokens: Token[] = [];
  lineStarts: number[] = [];

  constructor(readonly source: string) {
    let start = "";
    source.split(/\r?\n/g).forEach((line) => {
      this.lineStarts.push(start.length);
      start += line;
    });
  }

  getLocOfIndex(indexOfString: number): Location {
    const getLineOfIndex = (index: number) => {
      for (let i = 0; i < this.lineStarts.length - 1; i++) {
        if (this.lineStarts[i] < index && index < this.lineStarts[i + 1]) {
          return i;
        }
      }
      return last(this.lineStarts);
    };
    const line = getLineOfIndex(indexOfString);
    const column = indexOfString - this.lineStarts[line];
    return {
      line: line + 1,
      column: column - 1,
    };
  }

  scanTokens() {
    while (!this.isAtEnd()) {
      this.start = this.current;
      this.scanToken();
    }

    this.tokens.push({
      type: TokenType.EOF,
      lexeme: "",
      start: this.start,
      end: this.current,
      loc: {
        start: {
          line: this.line + 1,
          column: this.source.length - last(this.lineStarts) - 1,
        },
        end: {
          line: this.line + 1,
          column: this.source.length - last(this.lineStarts) - 1,
        },
      },
    });

    return this.tokens;
  }

  scanToken() {
    const char = this.advance();

    switch (char) {
      case "(":
        this.addToken(TokenType.LEFT_PAREN);
        break;
      case ")":
        this.addToken(TokenType.RIGHT_PAREN);
        break;
      case "{":
        this.addToken(TokenType.LEFT_BRACE);
        break;
      case "}":
        this.addToken(TokenType.RIGHT_BRACE);
        break;
      case "[":
        this.addToken(TokenType.LEFT_BRACKET);
        break;
      case "]":
        this.addToken(TokenType.RIGHT_BRACKET);
        break;
      case ",":
        this.addToken(TokenType.COMMA);
        break;
      case ".":
        this.addToken(TokenType.DOT);
        break;
      case "-":
        this.addToken(TokenType.MINUS);
        break;
      case "+":
        this.addToken(TokenType.PLUS);
        break;
      case ";":
        this.addToken(TokenType.SEMICOLON);
        break;
      case "*":
        this.addToken(TokenType.STAR);
        break;
      case "?":
        this.addToken(TokenType.QUESTION_MARK);
        break;
      case ":":
        this.addToken(TokenType.COLON);
        break;

      case "!":
        this.addToken(this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG);
        break;
      case "=":
        this.addToken(
          this.match("=") ? TokenType.EQUAL_EQUAL : TokenType.EQUAL
        );
        break;
      case "<":
        this.addToken(this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS);
        break;
      case ">":
        this.addToken(
          this.match("=") ? TokenType.GREATER_EQUAL : TokenType.GREATER
        );
        break;

      case "/":
        // TODO: support triple slash comments
        if (this.match("/")) {
          // A comment goes until the end of the line.
          while (this.peek() != "\n" && !this.isAtEnd()) this.advance();
          this.addToken(
            TokenType.COMMENT,
            this.source.slice(this.start + 2, this.current)
          );
        } else if (this.match("*")) {
          while (
            !(this.peek() === "*" && this.source[this.current + 1] === "/") &&
            !this.isAtEnd()
          )
            this.advance();
          this.advance();
          this.advance();
          this.addToken(
            TokenType.COMMENT,
            this.source.slice(this.start + 2, this.current - 2)
          );
        } else {
          this.addToken(TokenType.SLASH);
        }
        break;

      case " ":
      case "\r":
      case "\t":
        // Ignore whitespace.
        break;

      case "\n":
        this.line++;
        break;

      case `'`:
      case `"`:
        this.string(char);
        break;

      default:
        if (this.isDigit(char)) {
          this.number();
        } else if (this.isAlpha(char)) {
          this.identifier();
        } else {
          this.error(this.line, `unexpected char ${char}`);
        }
    }
  }

  private match(expected: string) {
    if (this.isAtEnd()) return false;
    if (this.source.charAt(this.current) != expected) return false;

    this.current++;
    return true;
  }

  private peek() {
    if (this.isAtEnd()) return "\0";
    return this.source.charAt(this.current);
  }

  advance() {
    this.current++;
    return this.source[this.current - 1];
  }

  addToken(type: TokenType, literal?: unknown) {
    this.tokens.push({
      type,
      literal,
      lexeme: this.source.slice(this.start, this.current),
      start: this.start,
      end: this.current,
      loc: {
        start: this.getLocOfIndex(this.start),
        end: this.getLocOfIndex(this.current),
      },
    });
  }

  private isAtEnd() {
    return this.current >= this.source.length;
  }

  private error(line: number, message: string): never {
    throw new Error(`line ${line}: ${message}`);
  }

  private string(quote: `"` | `'`) {
    while (this.peek() !== quote && !this.isAtEnd()) {
      this.advance();
    }

    if (this.isAtEnd() || this.peek() === "\n") {
      this.error(this.line, "Unterminated string.");
      return;
    }

    // The closing ".
    this.advance();

    const value = this.source.slice(this.start + 1, this.current - 1);
    this.addToken(TokenType.STRING, value);
  }

  private number() {
    const str = this.source.slice(this.start);
    let matched: RegExpMatchArray | null;
    let val: number;
    if ((matched = str.match(base16Re))) {
      val = Number(matched[0]);
    } else if ((matched = str.match(base8Re))) {
      val = parseInt(matched[0], 8);
    } else if ((matched = str.match(numberRe))) {
      val = Number(matched[0]);
    } else if ((matched = str.match(base10Re))) {
      val = Number(matched[0]);
    } else {
      throw new Error("should not happend");
    }

    this.current = this.start + matched[0].length;
    this.addToken(TokenType.NUMBER, val);
  }

  private identifier() {
    const str = this.source.slice(this.start);
    const matched = str.match(nameRe);
    if (!matched) {
      this.error(this.line, "should have matched id");
    }
    this.current = this.start + matched[0].length;

    if (PrimitiveTypes.includes(matched[0] as any)) {
      this.addToken(TokenType.PRIMITIVE_TYPE, matched[0]);
    } else {
      this.addToken(TokenType.IDENTIFIER, matched[0]);
    }
  }

  private isAlpha(c: string) {
    return /[a-zA-Z_]/.test(c);
  }

  private isDigit(c: string) {
    return c >= "0" && c <= "9";
  }
}
