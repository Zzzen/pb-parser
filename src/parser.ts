import { Scanner, Token, TokenType, Keyword } from "./scanner";
import {
  Comment,
  SyntaxStatement,
  ProtoFile,
  TopLevelDirective,
  ImportStatement,
  Package,
  FullIdentifier,
  Option,
  OptionName,
  Constant,
  TopLevelDef,
  Message,
  Reserved,
  MessageBody,
  MapField,
  ValueType,
} from "./ast";
import { last } from "./util";

export class Parser {
  pendingComments: Comment[] = [];
  current = 0;
  tokens: Token[] = [];

  constructor(private readonly source: string) {}

  handleComments() {
    while (this.check(TokenType.COMMENT)) {
      const token = this.peek();
      this.pendingComments.push({
        type: "Comment",
        value: token.literal as string,
        start: token.start,
        end: token.end,
      });
    }
  }

  parse() {
    this.tokens = new Scanner(this.source).scanTokens();
    return this.parseRoot();
  }

  parseSyntax(): SyntaxStatement | undefined {
    this.handleComments();
    if (this.checkKeyword(Keyword.SYNTAX)) {
      const syntax = this.advance();
      this.handleComments();
      this.consume(TokenType.EQUAL);
      const proto = this.consume(TokenType.STRING);
      const semicolon = this.consume(TokenType.SEMICOLON);
      return {
        type: "SyntaxStatement",
        value: proto.literal as "proto2" | "proto3",
        start: syntax.start,
        end: semicolon.end,
      };
    }
  }

  parseTopLevelDirectives(): TopLevelDirective[] {
    const topLevelDirectives: TopLevelDirective[] = [];
    while (!this.isAtEnd()) {
      if (this.checkKeyword(Keyword.IMPORT)) {
        topLevelDirectives.push(this.parseImport());
        continue;
      } else if (this.checkKeyword(Keyword.PACKAGE)) {
        topLevelDirectives.push(this.parsePackage());
        continue;
      } else if (this.checkKeyword(Keyword.OPTION)) {
        topLevelDirectives.push(this.parseOption());
        continue;
      } else if (this.match(TokenType.SEMICOLON)) {
        continue;
      } else if (
        this.peek().type === TokenType.IDENTIFIER &&
        [
          Keyword.MESSAGE,
          Keyword.ENUM,
          Keyword.EXTEND,
          Keyword.SERVICE,
        ].includes(this.peek().lexeme as any)
      ) {
        topLevelDirectives.push(this.parseTopLevelDef());
        continue;
      }

      this.error(this.peek(), `unexpected token`);
    }

    return topLevelDirectives;
  }

  /**
   * topLevelDef = message | enum | extend | service
   */
  parseTopLevelDef(): TopLevelDef {
    if (this.checkKeyword(Keyword.MESSAGE)) {
      return this.parseMessage();
    }

    this.error(this.peek(), "");
  }

  /**
   * message = "message" messageName messageBody
   * messageBody = "{" { field | enum | message | extend | extensions | group |
   *   option | oneof | mapField | reserved | emptyStatement } "}"
   */
  parseMessage(): Message {
    const messageToken = this.consumeKeyword(Keyword.MESSAGE);
    const name = this.consume(TokenType.IDENTIFIER).lexeme as string;
    this.consume(TokenType.LEFT_BRACE);
    const body = this.parseMessageBody();
    const end = this.consume(TokenType.RIGHT_BRACE);

    return {
      type: "Message",
      start: messageToken.start,
      end: end.end,
      body,
      name,
    };
  }

  parseMessageBody(): MessageBody[] {
    const body: MessageBody[] = [];
    while (!this.check(TokenType.RIGHT_BRACE)) {
      if (this.checkKeyword(Keyword.RESERVED)) {
        body.push(this.parserReserved());
        continue;
      } else if (this.checkKeyword(Keyword.MAP)) {
        body.push(this.parseMapField());
        continue;
      } else if (this.checkKeyword(Keyword.OPTION)) {
        body.push(this.parseOption());
        continue;
      }

      // emptyStatement
      if (this.match(TokenType.SEMICOLON)) {
        continue;
      }

      this.error(this.peek());
    }
    return body;
  }

  /**
   * mapField = "map" "<" keyType "," type ">" mapName "=" fieldNumber [ "[" fieldOptions "]" ] ";"
   * keyType = "int32" | "int64" | "uint32" | "uint64" | "sint32" | "sint64" |
   *          "fixed32" | "fixed64" | "sfixed32" | "sfixed64" | "bool" | "string"
   */

  parseMapField(): MapField {
    const start = this.consumeKeyword(Keyword.MAP);
    this.consume(TokenType.LESS);
    // keyType
    const keyToken = this.consume(TokenType.PRIMITIVE_TYPE);
    this.consume(TokenType.COMMA);

    const value = this.parseValueType();

    this.consume(TokenType.GREATER);

    const id = this.consume(TokenType.IDENTIFIER);

    this.consume(TokenType.EQUAL);

    const numberToken = this.consume(TokenType.NUMBER);

    let options: Option[] = [];
    if (this.match(TokenType.LEFT_BRACKET)) {
      options = this.parseFieldOptions();
      this.consume(TokenType.RIGHT_BRACKET);
    }

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "MapField",
      start: start.start,
      end: end.end,
      valueType: value,
      keyType: keyToken.lexeme,
      mapName: id.lexeme,
      fieldNumber: Number(numberToken.lexeme),
      options,
    };
  }

  /**
   * type = "double" | "float" | "int32" | "int64" | "uint32" | "uint64"
   *   | "sint32" | "sint64" | "fixed32" | "fixed64" | "sfixed32" | "sfixed64"
   *   | "bool" | "string" | "bytes" | messageType | enumType
   *
   *
   * messageType = [ "." ] { ident "." } messageName
   * enumType = [ "." ] { ident "." } enumName
   */
  parseValueType(): ValueType {
    if (this.match(TokenType.PRIMITIVE_TYPE)) {
      const token = this.previous();
      return {
        type: "ValueType",
        start: token.start,
        end: token.end,
        value: token.lexeme,
      };
    } else {
      const leadingDot = this.check(TokenType.DOT) ? this.advance() : undefined;
      const idents = [] as Token[];
      while (this.check(TokenType.IDENTIFIER)) {
        idents.push(this.advance());
        if (!this.match(TokenType.DOT)) {
          break;
        }
      }

      return {
        type: "ValueType",
        start: leadingDot ? leadingDot.start : idents[0].start,
        end: last(idents).end,
        value:
          (leadingDot ? "." : "") + idents.map((id) => id.lexeme).join("."),
      };
    }
  }

  /**
   * reserved = "reserved" ( ranges | fieldNames ) ";"
   * ranges = range { "," range }
   * range =  intLit [ "to" ( intLit | "max" ) ]
   * fieldNames = fieldName { "," fieldName }
   */
  parserReserved(): Reserved {
    const reserved = this.consumeKeyword(Keyword.RESERVED);
    if (this.check(TokenType.STRING)) {
      // fieldNames
      const ids = [this.advance()];
      while (this.match(TokenType.COMMA)) {
        ids.push(this.consume(TokenType.STRING));
      }

      const end = this.consume(TokenType.SEMICOLON);
      return {
        type: "Reserved",
        ranges: undefined,
        fieldName: ids.map((x) => x.literal as string),
        start: reserved.start,
        end: end.end,
      };
    } else {
      // ranges
      const ranges = [this.parseRange()];
      while (this.match(TokenType.COMMA)) {
        ranges.push(this.parseRange());
      }

      const end = this.consume(TokenType.SEMICOLON);
      return {
        type: "Reserved",
        ranges: ranges,
        fieldName: undefined,
        start: reserved.start,
        end: end.end,
      };
    }
  }

  parseRange(): [number, number | "max"] | [number] {
    const lit = this.consume(TokenType.NUMBER);
    const value = lit.literal as number;
    if (this.matchKeyword(Keyword.TO)) {
      if (this.matchKeyword(Keyword.MAX)) {
        return [value, "max"];
      } else {
        const toValue = this.consume(TokenType.NUMBER).literal as number;
        return [value, toValue];
      }
    } else {
      return [value];
    }
  }

  /**
   * option = "option" optionName  "=" constant ";"
   */
  parseOption(): Option {
    const start = this.consumeKeyword(Keyword.OPTION);

    const optionName = this.parseOptionName();

    this.consume(TokenType.EQUAL);

    const constant = this.parseConstant();

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "Option",
      name: optionName,
      value: constant,
      start: start.start,
      end: end.end,
    };
  }

  /**
   * fieldOptions = fieldOption { ","  fieldOption }
   */
  parseFieldOptions() {
    const opts = [this.parseFieldOption()];
    while (this.match(TokenType.COMMA)) {
      opts.push(this.parseFieldOption());
    }

    return opts;
  }

  /**
   * fieldOption = optionName "=" constant
   */
  parseFieldOption(): Option {
    const name = this.parseOptionName();
    this.consume(TokenType.EQUAL);
    const constant = this.parseConstant();
    return {
      type: "Option",
      start: name.start,
      end: constant.end,
      value: constant,
      name: name,
    };
  }

  /**
   * optionName = ( ident | "(" fullIdent ")" ) { "." ident }
   */
  parseOptionName(): OptionName {
    let name = "";
    const start = this.peek();
    if (this.check(TokenType.IDENTIFIER)) {
      name = this.consume(TokenType.IDENTIFIER).lexeme;
    } else {
      this.consume(TokenType.LEFT_PAREN);
      name += "(";

      const fullId = this.parseFullIdentifier();
      name += fullId.value;

      this.consume(TokenType.RIGHT_PAREN);
      name += ")";
    }

    while (this.match(TokenType.DOT)) {
      name += ".";
      name += this.consume(TokenType.IDENTIFIER).lexeme;
    }

    return {
      type: "OptionName",
      value: name,
      start: start.start,
      end: this.previous().end,
    };
  }

  /**
   * constant = fullIdent | ( [ "-" | "+" ] intLit ) | ( [ "-" | "+" ] floatLit ) |
   *                 strLit | boolLit
   */
  parseConstant(): Constant {
    if (this.check(TokenType.IDENTIFIER)) {
      return this.parseFullIdentifier();
    } else {
      let symbol: Token;
      if (this.check(TokenType.MINUS) || this.check(TokenType.PLUS)) {
        symbol = this.advance();
        const isMinue = symbol.type === TokenType.MINUS;
        return {
          type: "Literal",
          value:
            (isMinue ? -1 : +1) *
            (this.consume(TokenType.NUMBER).literal as number),
          start: symbol.start,
          end: symbol.end,
        };
      } else {
        const token = this.consume(TokenType.STRING);
        return {
          type: "Literal",
          value: token.literal as string,
          start: token.start,
          end: token.end,
        };
      }
    }
  }

  parsePackage(): Package {
    const packageToken = this.consumeKeyword(Keyword.PACKAGE);

    const name = this.parseFullIdentifier();

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "Package",
      name,
      start: packageToken.start,
      end: end.end,
    };
  }

  parseFullIdentifier(): FullIdentifier {
    const ids = [this.consume(TokenType.IDENTIFIER)];
    while (this.match(TokenType.DOT)) {
      ids.push(this.consume(TokenType.IDENTIFIER));
    }

    const fullId = ids.map((id) => id.lexeme).join(".");

    return {
      type: "FullIdentifier",
      value: fullId,
      start: ids[0].start,
      end: ids[ids.length - 1].end,
    };
  }

  parseImport(): ImportStatement {
    const importToken = this.consumeKeyword(Keyword.IMPORT);
    let modifier: Token | undefined;
    if (this.checkKeyword(Keyword.WEAK) || this.checkKeyword(Keyword.PUBLIC)) {
      modifier = this.advance();
    }

    const strLit = this.consume(TokenType.STRING);

    const end = this.consume(TokenType.SEMICOLON);

    return {
      type: "ImportStatement",
      file: strLit.literal as string,
      modifier: modifier?.lexeme as "weak" | "public",
      start: importToken.start,
      end: end.end,
    };
  }

  /**
   * proto = syntax { import | package | option | topLevelDef | emptyStatement }
   */
  parseRoot(): ProtoFile {
    const syntax = this.parseSyntax();
    const body = this.parseTopLevelDirectives();
    return {
      type: "ProtoFile",
      syntax,
      body,
      start: this.tokens[0].start,
      end: this.tokens[this.tokens.length - 1].end,
    };
  }

  private previous() {
    return this.tokens[this.current - 1];
  }

  check(type: TokenType) {
    if (this.isAtEnd()) return false;
    return this.peek().type == type;
  }

  checkKeyword(type: Keyword) {
    if (this.isAtEnd()) return false;
    return (
      this.peek().type == TokenType.IDENTIFIER && this.peek().lexeme === type
    );
  }

  checkNext(type: TokenType) {
    return this.tokens[this.current + 1]?.type === type;
  }

  match(...types: TokenType[]) {
    for (const type of types) {
      if (this.check(type)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  matchKeyword(...types: Keyword[]) {
    for (const type of types) {
      if (this.checkKeyword(type)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  private consume(type: TokenType) {
    if (this.check(type)) return this.advance();

    const current = this.tokens[this.current];
    const errorMessage = `expected ${type}, but got ${current.type} at line ${current.line}`;

    throw this.error(this.peek(), errorMessage);
  }

  private consumeKeyword(type: Keyword) {
    if (this.checkKeyword(type)) return this.advance();

    const current = this.tokens[this.current];
    const errorMessage = `expected ${type}, but got ${current.type} at line ${current.line}`;

    throw this.error(this.peek(), errorMessage);
  }

  private error(token: Token, message?: string): never {
    throw new Error(
      `parsing Error: unexpect token ${token.type}(${token.lexeme}), ${message}`
    );
  }

  private advance() {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private isAtEnd() {
    return this.peek().type === TokenType.EOF;
  }

  private peek() {
    return this.tokens[this.current];
  }
}
