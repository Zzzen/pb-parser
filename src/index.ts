import { Parser } from "./parser";

export function parse(src: string, _options: unknown) {
  return new Parser(src).parse();
}

export * from "./parser";
export * from "./scanner";
export * from "./ast";
