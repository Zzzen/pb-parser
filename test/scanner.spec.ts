import { Scanner, TokenType, Token } from "../src/scanner";

function scan(src: string) {
  return new Scanner(src).scanTokens();
}

describe("Scanner", () => {
  it("scan combinations of simple symbols", () => {
    expect(scan("{}")).toMatchSnapshot();
    expect(scan("[]")).toMatchSnapshot();
  });

  it("scan comments", () => {
    const lineComments = `
//comment1
// comment2
`;
    expect(scan(lineComments)).toMatchSnapshot();

    const blockComments = `
/**
 * comment1
 * comment2
 */
        `;

    expect(scan(blockComments)).toMatchSnapshot();
  });

  it("scan numbers", () => {
    const code = `
23
234.35
0x43
0345
1e3
        `;
    expect(scan(code)).toMatchSnapshot();
  });

  it("scan identifiers", () => {
    const code = `
foo
bar
_foo
package foo;
import
        `;

    expect(scan(code)).toMatchSnapshot();
  });
});
