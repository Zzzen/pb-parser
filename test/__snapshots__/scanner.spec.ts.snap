// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scanner scan combinations of simple symbols 1`] = `
Array [
  Object {
    "end": 1,
    "lexeme": "{",
    "line": 0,
    "literal": undefined,
    "start": 0,
    "type": "LEFT_BRACE",
  },
  Object {
    "end": 2,
    "lexeme": "}",
    "line": 0,
    "literal": undefined,
    "start": 1,
    "type": "RIGHT_BRACE",
  },
  Object {
    "end": 2,
    "lexeme": "",
    "line": 0,
    "start": 1,
    "type": "EOF",
  },
]
`;

exports[`Scanner scan combinations of simple symbols 2`] = `
Array [
  Object {
    "end": 1,
    "lexeme": "[",
    "line": 0,
    "literal": undefined,
    "start": 0,
    "type": "LEFT_BRACKET",
  },
  Object {
    "end": 2,
    "lexeme": "]",
    "line": 0,
    "literal": undefined,
    "start": 1,
    "type": "RIGHT_BRACKET",
  },
  Object {
    "end": 2,
    "lexeme": "",
    "line": 0,
    "start": 1,
    "type": "EOF",
  },
]
`;

exports[`Scanner scan comments 1`] = `
Array [
  Object {
    "end": 11,
    "lexeme": "//comment1",
    "line": 1,
    "literal": "comment1",
    "start": 1,
    "type": "COMMENT",
  },
  Object {
    "end": 23,
    "lexeme": "// comment2",
    "line": 2,
    "literal": " comment2",
    "start": 12,
    "type": "COMMENT",
  },
  Object {
    "end": 24,
    "lexeme": "",
    "line": 3,
    "start": 23,
    "type": "EOF",
  },
]
`;

exports[`Scanner scan comments 2`] = `
Array [
  Object {
    "end": 32,
    "lexeme": "/**
 * comment1
 * comment2
 */",
    "line": 1,
    "literal": "*
 * comment1
 * comment2
 ",
    "start": 1,
    "type": "COMMENT",
  },
  Object {
    "end": 41,
    "lexeme": "",
    "line": 2,
    "start": 40,
    "type": "EOF",
  },
]
`;

exports[`Scanner scan identifiers 1`] = `
Array [
  Object {
    "end": 4,
    "lexeme": "foo",
    "line": 1,
    "literal": "foo",
    "start": 1,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 8,
    "lexeme": "bar",
    "line": 2,
    "literal": "bar",
    "start": 5,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 13,
    "lexeme": "_foo",
    "line": 3,
    "literal": "_foo",
    "start": 9,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 21,
    "lexeme": "package",
    "line": 4,
    "literal": "package",
    "start": 14,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 25,
    "lexeme": "foo",
    "line": 4,
    "literal": "foo",
    "start": 22,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 26,
    "lexeme": ";",
    "line": 4,
    "literal": undefined,
    "start": 25,
    "type": "SEMICOLON",
  },
  Object {
    "end": 33,
    "lexeme": "import",
    "line": 5,
    "literal": "import",
    "start": 27,
    "type": "IDENTIFIER",
  },
  Object {
    "end": 42,
    "lexeme": "",
    "line": 6,
    "start": 41,
    "type": "EOF",
  },
]
`;

exports[`Scanner scan numbers 1`] = `
Array [
  Object {
    "end": 3,
    "lexeme": "23",
    "line": 1,
    "literal": 23,
    "start": 1,
    "type": "NUMBER",
  },
  Object {
    "end": 10,
    "lexeme": "234.35",
    "line": 2,
    "literal": 234.35,
    "start": 4,
    "type": "NUMBER",
  },
  Object {
    "end": 15,
    "lexeme": "0x43",
    "line": 3,
    "literal": 67,
    "start": 11,
    "type": "NUMBER",
  },
  Object {
    "end": 20,
    "lexeme": "0345",
    "line": 4,
    "literal": 229,
    "start": 16,
    "type": "NUMBER",
  },
  Object {
    "end": 24,
    "lexeme": "1e3",
    "line": 5,
    "literal": 1000,
    "start": 21,
    "type": "NUMBER",
  },
  Object {
    "end": 33,
    "lexeme": "",
    "line": 6,
    "start": 32,
    "type": "EOF",
  },
]
`;
